chapters/chapter_5/section_5.md

```diff
@@ -172,233 +318,157 @@ dummy_worker_insert(LogThreadedDestDriver *d, LogMessage *msg)
+  self->super.thread_deinit = _thread_deinit;
+  self->super.insert = _dw_insert;
+  self->super.free_fn = _dw_free;
+  self->super.connect = _connect;
+  self->super.disconnect = _disconnect;
 
-  return TRUE;
+  return &self->super;
 }
-
-const ModuleInfo module_info =
-{
-  .canonical_name = "dummy",
-  .version = SYSLOG_NG_VERSION,
-  .description = "This is a dummy destination for syslog-ng.",
-  .core_revision = SYSLOG_NG_SOURCE_REVISION,
-  .plugins = &dummy_plugin,
-  .plugins_len = 1,
-};
 ```
 
-This unit can be separated into 5 parts:
-
-1. configuration functions (setters)
-2. utility functions
-3. functions running in separate thread
-4. functions running in the main thread
-5. plugin-construction declarations
-
-Functions from category 2-4 are implementations of the LogThreadedDestDriver's virtual methods.
-In `dummy_dd_new`, we pass these function pointers to the base classes.
-
-Our example overrides these 6 virtual methods:
+Our example overrides these virtual methods:
+- `thread_init (_thread_init)`: if you need to initialize thread specific resources, you can do them here. We are saving the thread id in this example. Non-thread specific resources may be created in the constructor (`example_destination_dw_new`).
+- `thread_deinit (_thread_deinit)`: if you created resources during `thread_init`, you need to deallocate them here.
+- `free_fn (_dw_free)` destructor for resources created in `example_destination_dw_new`.
+- `insert (_dw_insert)`: It formats the received message and sends to an actual destination.
+- `connect (_dw_connect)`: This is called after `thread_init`, before the first `insert`, and each time you signal broken connection (returning `LTR_NOT_CONNECTED` from `insert`). You can open files or establish connections here.
+- `disconnect (_dw_disconnect)`: This is called before deinit, or when you signal broken connection from insert. You can close files or sockets here.
 
-- `init (dummy_dd_init)`: destination constructor
-- `free_fn (dummy_dd_free)`: destination destructor
-- `thread_init (dummy_worker_thread_init)`: worker thread initialization
-- `thread_deinit (dummy_worker_thread_deinit)`: worker thread deinitialization
-- `disconnect (dummy_dd_disconnect)`: destination disconnects (on error or drop)
+### grammar keywords
 
-The most important method is the `insert (dummy_worker_insert)`, where you can format the received message and send to an actual destination.
+`example_destination-parser.c` contains a list of available keywords that can be referred in the grammar. A keyword is an integer with a string representation. The integer is defined in `example_destination-grammar.ym`: see the example below. The string representation is defined in `example_destination-grammar.c`.
 
-### dummy-parser
+The example destination module will support two keywords: `example_destination` and `filename`. You need to replace `example_destination-keywords` in `example_destination-grammar.c` with:
```

Comment:
not grammar.c, but parser.c, right?

See at line 408 too:
> The string representation is defined in `example_destination-grammar.c`.
Here I assume it correct, as grammar.c file is indeed generated.

Reply:
Fixed thanks